### PR TITLE
Javalab: load and create code review comments

### PR DESCRIPTION
**Pull request by bencodeorg**
First steps of connecting back and front ends for Javalab code review. This PR includes:
- Gets code review comments from database to display for a given project.
- Allows creating new comments on a project.

## Testing story

Tested manually, and added unit tests to cover basic new behavior (adds a new comment to the list of visible comments on successful comment create).

## Follow-up work

_Follow-ups explicitly tied to this work:_

- Handle error cases in submitting comments (eg, fails to submit comment for some reason)
- Do we want a spinner or some sort of waiting UI for waiting for request to finish?
- Add support for project version in saving comments -- it looks like we don't have the project version as a standard part of `pageConstants`. Feels like something that may be worth having, but thought I could do that work separately. As is, we just render all comments for given project -- eventually, we'll want to log the project version with the comment so we can style the comment appropriately when we show it on the project. We gray out comments that were made on older versions.

_Follow-ups around code review_

- Adding UI for resolving and deleting comments, and hook up this UI to back end
- Permissioning around who sees what on front end (ie, when to show review tab)
- Styling around timestamps for when comments were made

### CODE REVIEW

File: apps/src/templates/instructions/codeReview/CommentEditor.jsx [hunk 1]
```diff
@@ -1,20 +1,33 @@
...
+    onNewCommentSubmit: PropTypes.func.isRequired,
+    token: PropTypes.string.isRequired
   };
 
-  commentChanged = event => {
-    this.setState({
-      comment: event.target.value,
-      isEditing: true
-    });
+  state = {comment: ''};
```
Comment (bencodeorg) [inline apps/src/templates/instructions/codeReview/CommentEditor.jsx:11]: Removing `isEditing` is an unrelated change that just makes it so when the comment is blank, the buttons disappear (instead of having a dedicated `isEditing` flag). Also a bit of CSS cleanup.

File: apps/src/templates/instructions/ReviewTab.jsx [hunk 2]
```diff
@@ -1,18 +1,30 @@
...
+    comments: PropTypes.arrayOf(commentShape).isRequired,
+    token: PropTypes.string.isRequired
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      readyForReview: false,
+      comments: this.props.comments
```
Comment (molly-moen) [inline apps/src/templates/instructions/ReviewTab.jsx]: In general we want to avoid copying props into state because it can cause issues (for example, in the future if another component wants to use this and thinks it can control the comments via props it won't work). We should either have an onChange method that is handled by `TopInstructions` or put the comments in redux.

Comment (bencodeorg) [reply apps/src/templates/instructions/ReviewTab.jsx]: > in the future if another component wants to use this and thinks it can control the comments via props it won't work).

For my understanding, can you clarify what you mean by this?

Comment (molly-moen) [reply apps/src/templates/instructions/ReviewTab.jsx]: if we have some component `SampleComponent` which has `ReviewTab` in it, and `SampleComponent` changes the `comments` props, `ReviewTab` will carry on with its copied state instead of using the new props (which is what we want here, but may not be what `SampleComponent` would expect). This is an [article](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html) for a more up-to-date version of react, but generally copying props to state is an anti-pattern. LMK if that made sense

Comment (bencodeorg) [reply apps/src/templates/instructions/ReviewTab.jsx]: Ok I think I'm getting it! It also mentions using a `key` as an option to force creating a new component instance (and all children) vs. updating an existing one, would that be ok, or is Redux preferable?

https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key

To test my understanding, am I correct that this component is in heavy violation of that anti-pattern?

https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/lib/levelbuilder/unit-editor/UnitEditor.jsx#L58

Comment (molly-moen) [reply apps/src/templates/instructions/ReviewTab.jsx]: I think either approach (key or redux) would be fine here. To my eyes that component does look in violation of the anti-pattern

Comment (jmkulwik) [reply apps/src/templates/instructions/ReviewTab.jsx]: One question that comes to mind: What is the source of truth for comments that should be displayed? For example, if we add a refresh button that pulls in new comments from the DB, how will the new comments be synced with the existing comments?

If I'm understanding correctly, TopInstructions is handling information flow from the DB whereas ReviewTab is handling information flow from the "submit" box. With information being handled in two different places, we're reaching for the anti-pattern of storing props in state.

One possible solution:

TopInstructions seems to be doing too much. It's handling a _lot_ of display logic as well as managing a lot of information for each of its child components. Instead, we can have TopInstructions only decide whether or not to display the review tab.

The review tab, if displayed, would in turn handle the DB call as well as updates from the submit box. It would manage the source-of-truth list of comments.

The review tab would then continue to have child components that display the comments and the submit box.

File: apps/src/templates/instructions/TopInstructions.jsx [hunk 3]
```diff
@@ -191,6 +194,21 @@ class TopInstructions extends Component {
           })
       );
     }
+
+    if (displayReviewTab) {
+      const channelId = getStore().getState().pageConstants.channelId;
+
+      promises.push(
+        topInstructionsDataApi
+          .getCodeReviewCommentsForProject(channelId, 'yyy')
```
Comment (jmkulwik) [inline apps/src/templates/instructions/TopInstructions.jsx]: Is 'yyy' here a temporary placeholder? (if so, mind adding a comment?)

File: apps/src/templates/instructions/codeReview/CommentEditor.jsx [hunk 4]
```diff
@@ -1,19 +1,35 @@
...
-  commentChanged = event => {
-    this.setState({
-      comment: event.target.value,
-      isEditing: true
+  state = {comment: ''};
+
+  commentChanged = event => this.setState({comment: event.target.value});
+
+  submitNewComment = () => {
+    $.ajax({
```
Comment (jmkulwik) [inline apps/src/templates/instructions/codeReview/CommentEditor.jsx]: Move the ajax call to ReviewTab or TopInstructions. 

Reasoning: CommentEditor is set up very similarly to a [controlled component](https://reactjs.org/docs/forms.html). This makes a lot of sense and we should lean into this and go ahead and make it a controlled component. Since it's set up as a controlled component, however, I was a surprised to see the ajax call here. I kept looking for it in TopInstructions (where the other ajax call is) or in ReviewTab (which controls the state of the comment list) and had assumed it hadn't been implemented yet. Moving the ajax call to one of those other locations will make the behavior a bit more expected for future maintainers.

File: apps/src/templates/instructions/topInstructionsDataApi.js [hunk 5]
```diff
@@ -37,3 +37,14 @@ export function incrementVisitCount(latestFeedbackId, token) {
     headers: {'X-CSRF-Token': token}
   });
 }
+
+export function getCodeReviewCommentsForProject(channelId, projectVersion) {
+  return $.ajax({
+    url: `/code_review_comments/project_comments`,
```
Comment (jmkulwik) [inline apps/src/templates/instructions/topInstructionsDataApi.js]: I'd recommend moving this and the other comments ajax call to their own helper library. This would keep all the code for codeReviews in the same place and keep it isolated from other functionality in topInstructions. This makes codeReview a bit more maintainable as changes to its code will have no effect on other portions of topInstructions. It'll also be more versatile as it won't be tied to functionality in topInstructions.

File: apps/src/templates/instructions/ReviewTab.jsx (hunk 2 above)
Comment (bencodeorg) [reply apps/src/templates/instructions/ReviewTab.jsx]: That sounds reasonable, as well as your other suggestions :). Let me take another pass and see if I can clean this up.

Comment (bencodeorg) [reply apps/src/templates/instructions/ReviewTab.jsx]: Two questions after looking at this briefly:
1. [We remove `ReviewTab` from the DOM when we switch away from the tab](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/templates/instructions/TopInstructions.jsx#L693), so if I move the AJAX call into the review tab (in, say, its own `componentDidMount`, we make a new request each time you switch to the tab). Maybe this is a positive from a refresh button perspective?
2. Just confirming that this feels worthwhile given how similar this feature is to teacher feedback to implement it in a different (even if better) way? May result in some context switching to understand we use different patterns to implement such closely related features. I think maybe the answer is yes (they're not entirely similar), but just want to make sure.

Comment (bencodeorg) [reply apps/src/templates/instructions/ReviewTab.jsx]: Maybe easiest to just hide `ReviewTab` when it's not selected like we do here

https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/templates/instructions/TopInstructions.jsx#L700

Comment (bencodeorg) [reply apps/src/templates/instructions/ReviewTab.jsx]: display: none seems like a frowned upon pattern as well :). Third idea (Molly's idea): store comments in Redux so we don't need to make network requests any time someone switches to the review tab.

Comment (jmkulwik) [reply apps/src/templates/instructions/ReviewTab.jsx]: > display: none seems like a frowned upon pattern as well

^ TIL :) What's the reasoning for that?

Comment (jmkulwik) [reply apps/src/templates/instructions/ReviewTab.jsx]: > so if I move the AJAX call into the review tab (in, say, its own componentDidMount, we make a new request each time you switch to the tab). Maybe this is a positive from a refresh button perspective?

^ That's a good point. If you don't mind, I'm going to schedule a message to Hannah and Jamila in the AM to see if that would be preferable.

Comment (jmkulwik) [reply apps/src/templates/instructions/ReviewTab.jsx]: > Just confirming that this feels worthwhile given how similar this feature is to teacher feedback to implement it in a different (even if better) way?

^ I'd say so. A possible mitigation is to leave a comment indicating which pattern people should use going forward. That may also give a future engineer an easy blueprint to follow if they want to do a little bit of refactoring of TopInstructions.

Comment (jmkulwik) [reply apps/src/templates/instructions/ReviewTab.jsx]: One final possibility (redux is also a great idea): [LibraryManagerDialog](https://github.com/code-dot-org/code-dot-org/blob/d86cd51e40578bf02853cd377503e7c3507f49bb/apps/src/code-studio/components/libraries/LibraryManagerDialog.jsx#L156) ran into a similar issue. The pattern followed there went like this:

LibraryManagerDialog mounts on page load but doesn't display anything.

Its parent indicates to LibraryManagerDialog whether it should be "open" or "closed"

When LibraryManagerDialog opens, it displays child components such as libraryListItem and triggers an ajax query. When the query is finished, LibraryManagerDialog updates its children to display the result of the ajax query. It also updates its own internal state.

When LibraryManagerDialog closes, it unmounts all of its child components, so nothing is displayed anymore. But LibraryManagerDialog itself is still mounted, so the state is preserved. Next time it is opened, the dialog is populated with the state maintained from the previous ajax query. The query will run again to take any updates from the server, but the user can start interacting with the UI immediately.

This may be too similar to the `display: none` scenario you mentioned earlier - definitely worth understanding whether this also violates the same principles.

Comment (bencodeorg) [reply apps/src/templates/instructions/ReviewTab.jsx]: re: avoiding display: none, maybe totally frowned upon was a bit of an exaggeration, but I was just reading opinions about how you can pollute the DOM with too much hidden content (maybe not a huge problem in this case).

That said, I think using display: none here or something like the LibraryManagerDialog approach you mentioned would result in the review tab being in the DOM everywhere that we show TopInstructions, which does feel like a code smell?

I have a couple of new commits that I'll share that moves both AJAX requests into `ReviewTab`. It is much cleaner I think, but does re-request on each click of the Review tab, which I can try and fix via Redux, or do as a follow-up.

File: apps/src/templates/instructions/TopInstructions.jsx [hunk 6]
```diff
@@ -191,6 +191,7 @@ class TopInstructions extends Component {
           })
       );
     }
+
```
Comment (molly-moen) [inline apps/src/templates/instructions/TopInstructions.jsx:194]: nit: remove this blank line?

File: apps/src/templates/instructions/codeReview/CommentEditor.jsx [hunk 7]
```diff
@@ -28,25 +25,23 @@ export default class CommentEditor extends Component {
...
         <div
           style={{
-            ...styles.container,
-            ...styles.footer,
-            justifyContent: 'flex-end'
+            ...styles.buttonContainer
           }}
         >
-          {this.state.isEditing && (
+          {comment && (
```
Comment (molly-moen) [inline apps/src/templates/instructions/codeReview/CommentEditor.jsx:31]: nit: can you do `comment && (both buttons)` or do you then need to surround both buttons with another tag?

File: apps/src/templates/instructions/ReviewTab.jsx [hunk 8]
```diff
@@ -1,22 +1,54 @@
...
+    forceRecreateEditorKey: 0
+  };
+
+  componentDidMount() {
+    const channelId = getStore().getState().pageConstants.channelId;
+
+    // projectVersion = 'latest' is a placeholder until we implement
+    // storing project version when saving comments
+    codeReviewDataApi
+      .getCodeReviewCommentsForProject(channelId, 'latest')
```
Comment (molly-moen) [inline apps/src/templates/instructions/ReviewTab.jsx]: could using 'latest' potentially cause issues when we do store project version with comments (ex we are still querying for latest but now all comments have versions so we don't get any results)?

File: apps/src/templates/instructions/codeReview/CommentEditor.jsx (hunk 7 above)
Comment (bencodeorg) [reply apps/src/templates/instructions/codeReview/CommentEditor.jsx:31]: I tried but got yelled at by our version of React 😬

File: apps/src/templates/instructions/TopInstructions.jsx (hunk 6 above)
Comment (bencodeorg) [reply apps/src/templates/instructions/TopInstructions.jsx:194]: I guess no longer related to this PR to have changes to this file, but there are spaces between each if block so it seems more consistent 🤷

File: apps/src/templates/instructions/ReviewTab.jsx (hunk 8 above)
Comment (bencodeorg) [reply apps/src/templates/instructions/ReviewTab.jsx]: I was assuming we had to finish the project version stuff before the pilot, I meant short term placeholder

Comment (molly-moen) [reply apps/src/templates/instructions/ReviewTab.jsx]: I think our plan for pilot is to show comments for all versions, is that right @jmkulwik?

Comment (bencodeorg) [reply apps/src/templates/instructions/ReviewTab.jsx]: Ooh interesting, my understanding was we show all comments on all versions of a given project, just style comments on older versions differently. In any case, if this isn't a feature we 100% need by the pilot, I could make project version nullable (it currently is not, as I thought it was one of the primary identifiers for a comment)

Comment (molly-moen) [reply apps/src/templates/instructions/ReviewTab.jsx]: yep your understanding is correct, but we won't be doing that pre-pilot. Why do we want to query comments by version since we always want them all anyway?

Comment (bencodeorg) [reply apps/src/templates/instructions/ReviewTab.jsx]: 🤦 totally don't need to query using it :) I'll remove the code where we're querying with it, and modify the AJAX calls to only include the channelId (and just set up the controller to put in a temporary placeholder value). I'll write a migration to make the field nullable, and remove the temporary placeholder version string being used in the controller at that point.

**Review by molly-moen — APPROVED** [review]
Nice!
